### PR TITLE
Fix for https://github.com/aspnet/Mvc/issues/4903.

### PR DESCRIPTION
**Pull request by Haplois**
Summary of the changes
 - Implemented support for placeholder in `<textarea>` elements.

Addresses aspnet/Mvc#4903

### CODE REVIEW

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/DefaultHtmlGenerator.cs [hunk 1]
```diff
@@ -1289,6 +1286,27 @@ internal static object GetModelStateValue(ViewContext viewContext, string key, T
         }
 
         /// <summary>
+        /// Adds placeholder attribute to the <paramref name="tagBuilder" />.
```
Comment (Eilon) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/DefaultHtmlGenerator.cs]: `Adds *a* placeholder ...`

**Comment by Eilon** [comment]
@Haplois thanks for the PR! Could you add a unit test to cover this new case? Perhaps there's already some existing tests that can be copy/pasted to add this scenario?

@dougbu the code overall looks good to me - can you review? Not sure if any null checks are needed.

**Comment by Haplois** [comment]
@Eilon, I'll add appropriate tests today (GMT +02 :)).

**Comment by Haplois** [comment]
@Eilon tests are added.

File: test/Microsoft.AspNetCore.Mvc.ViewFeatures.Test/Rendering/HtmlHelperTextBoxAreaTest.cs [hunk 2]
```diff
@@ -0,0 +1,50 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using Microsoft.AspNetCore.Mvc.TestCommon;
+using System.ComponentModel.DataAnnotations;
+using Xunit;
+
+namespace Microsoft.AspNetCore.Mvc.Rendering
+{
+    public class HtmlHelperTextBoxAreaTest
```
Comment (dougbu) [inline test/Microsoft.AspNetCore.Mvc.ViewFeatures.Test/Rendering/HtmlHelperTextBoxAreaTest.cs]: Rename class to `HtmlHelperTextAreaTest` and then make filename match.

File: test/Microsoft.AspNetCore.Mvc.ViewFeatures.Test/Rendering/HtmlHelperTextBoxAreaTest.cs [hunk 3]
```diff
@@ -0,0 +1,50 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using Microsoft.AspNetCore.Mvc.TestCommon;
+using System.ComponentModel.DataAnnotations;
+using Xunit;
```
Comment (dougbu) [inline test/Microsoft.AspNetCore.Mvc.ViewFeatures.Test/Rendering/HtmlHelperTextBoxAreaTest.cs]: Sort w/ `System` namespaces first.

File: test/Microsoft.AspNetCore.Mvc.ViewFeatures.Test/Rendering/HtmlHelperTextBoxAreaTest.cs [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+        public void TextAreaFor_DoesNotGeneratePlaceholderAttribute_WhenNoPlaceholderPresentInModel()
+        {
+            // Arrange            
+            var model = new TextAreaModelWithoutAPlaceholder();
+            var helper = DefaultTemplatesUtilities.GetHtmlHelper(model);
+
+            // Act
+            var result = HtmlContentUtilities.HtmlContentToString(helper.TextAreaFor(m => m.Property1));
+            
+            Assert.False(result.Contains(@"placeholder=""HtmlEncode[[placeholder]]"""));
```
Comment (dougbu) [inline test/Microsoft.AspNetCore.Mvc.ViewFeatures.Test/Rendering/HtmlHelperTextBoxAreaTest.cs]: Add `// Assert`

File: test/Microsoft.AspNetCore.Mvc.ViewFeatures.Test/Rendering/HtmlHelperTextBoxAreaTest.cs [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+    {
+        [Fact]
+        public void TextAreaFor_GeneratesPlaceholderAttribute_WhenDisplayAttributePromptIsSetAndTypeIsValid()
+        {
+            // Arrange            
+            var model = new TextAreaModelWithAPlaceholder();
+            var helper = DefaultTemplatesUtilities.GetHtmlHelper(model);
+
+            // Act
+            var result = HtmlContentUtilities.HtmlContentToString(helper.TextAreaFor(m => m.Property1));
```
Comment (dougbu) [inline test/Microsoft.AspNetCore.Mvc.ViewFeatures.Test/Rendering/HtmlHelperTextBoxAreaTest.cs]: Not major i.e. mostly for future reference: Better to break up expressions to make it clear what the real `// Act` is. Here that would mean `var result = helper.TextAreaFor(m => m.Property1);` and doing the `HtmlContentToString()` conversion under `// Assert`.

File: test/Microsoft.AspNetCore.Mvc.ViewFeatures.Test/Rendering/HtmlHelperTextBoxAreaTest.cs [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+
+            // Assert 
+            Assert.True(result.Contains(@"placeholder=""HtmlEncode[[placeholder]]"""));
```
Comment (dougbu) [inline test/Microsoft.AspNetCore.Mvc.ViewFeatures.Test/Rendering/HtmlHelperTextBoxAreaTest.cs]: Instead use `Assert.Contains()` but add `StringComparison.Ordinal`. Also use `Assert.DoesNotContains()` in test below.

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/DefaultHtmlGenerator.cs [hunk 7]
```diff
@@ -1289,6 +1286,27 @@ internal static object GetModelStateValue(ViewContext viewContext, string key, T
...
 
         /// <summary>
+        /// Adds a placeholder attribute to the <paramref name="tagBuilder" />.
+        /// </summary>
+        /// <param name="viewContext">A <see cref="ViewContext"/> instance for the current scope.</param>
+        /// <param name="tagBuilder">A <see cref="TagBuilder"/> instance.</param>
+        /// <param name="modelExplorer">The <see cref="ModelExplorer"/> for the <paramref name="expression"/>.</param>
+        /// <param name="expression">Expression name, relative to the current model.</param>
+        protected virtual void AddPlaceholderAttribute(
+            ViewContext viewContext,
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/DefaultHtmlGenerator.cs]: Suggest passing in just the `ViewDataDictionary` unless you feel an override is likely to use other `ViewContext` properties. Consistent parameters with `AddValidationAttributes()` isn't a goal.

**Comment by dougbu** [comment]
:wait: mostly for a bit of test cleanup

**Comment by Haplois** [comment]
@dougbu I modified the source.

**Comment by dougbu** [comment]
:shipit: thanx very much @Haplois!

**Comment by dougbu** [comment]
a852352 

@Haplois additional thanks for your `HtmlHelperTextBoxTest` cleanup.

**Comment by Haplois** [comment]
@dougbu No worries, glad if I helped.

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/DefaultHtmlGenerator.cs [hunk 8]
```diff
@@ -1289,6 +1286,27 @@ internal static object GetModelStateValue(ViewContext viewContext, string key, T
...
         /// <summary>
+        /// Adds a placeholder attribute to the <paramref name="tagBuilder" />.
+        /// </summary>
+        /// <param name="viewData">A <see cref="ViewDataDictionary"/> instance for the current scope.</param>
+        /// <param name="tagBuilder">A <see cref="TagBuilder"/> instance.</param>
+        /// <param name="modelExplorer">The <see cref="ModelExplorer"/> for the <paramref name="expression"/>.</param>
+        /// <param name="expression">Expression name, relative to the current model.</param>
+        protected virtual void AddPlaceholderAttribute(
+            ViewDataDictionary viewData,
+            TagBuilder tagBuilder,
```
Comment (Haplois) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/DefaultHtmlGenerator.cs]: &ğxxhttps://technet.microsoft.com/en-us/sysinternals/ee656415.aspxkjtc

**Pull request merged**